File: four-connect.py
import sys, os


ROWS = 6
COLS = 7
arr = [["0"] * COLS for _ in range(ROWS)]


def print_board():
    for r in range(ROWS):
        for c in range(COLS):
            end_str = "\n" if (COLS - 1) == c else ""
            print(arr[r][c] + " ", end=end_str)


def place_coin(col, coin):
    row = 0
    if arr[row][col] != "0":
        print("Column is full, try again")
        return False

    while arr[row][col] == "0":
        # place if the next row is the end or the next is one is != 0
        if row + 1 == ROWS or arr[row + 1][col] != "0":
            arr[row][col] = coin
            break
        row += 1
    return True


def check_draw():
    for r in range(ROWS):
        for c in range(COLS):
            if arr[r][c] == "0":
                return False
    return True


def is_out_of_bounds(row, col):
    return row < 0 or col < 0 or col >= COLS or row >= ROWS


def check_is_winner(coin):
    for r in range(ROWS):
        for c in range(COLS):
            if arr[r][c] == coin:
                directions = [
                    (1, 0),  # Right
                    (-1, 0),  # Left
                    (0, 1),  # Up
                    (0, -1),  # Down
                    (1, 1),  # Up-Right
                    (-1, 1),  # Up-Left
                    (1, -1),  # Down-Right
                    (-1, -1),  # Down-Left
                ]
                for dx, dy in directions:
                    newR = dx + r
                    newC = dy + c
                    found_coins = 1
                    for _ in range(3):
                        # if out of bounds, break
                        if is_out_of_bounds(newR, newC):
                            break
                        if arr[newR][newC] == coin:
                            found_coins += 1
                        newR += dx
                        newC += dy
                    if found_coins == 4:
                        return True

    return False


def play():
    current_turn = "R"
    while True:
        while True:
            os.system("cls")
            print(f"Its {current_turn} Turn:")
            column_input = ""
            while not column_input.isnumeric():
                print("pick a column to place the coin 0-5")
                print_board()
                column_input = input()
                if is_out_of_bounds(1, int(column_input)):
                    column_input = ""
                    os.system("cls")
                    print("Column is out of bounds please try again")
            column_input = int(column_input)
            is_placed = place_coin(column_input, current_turn)
            if is_placed:
                current_turn = "R" if current_turn == "Y" else "Y"
                break
        if check_is_winner("Y"):
            os.system("cls")
            print("Y won the game!")
            print_board()
            break
        if check_is_winner("R"):
            os.system("cls")
            print("R won the game!")
            print_board()
            break
        if check_draw():
            os.system("cls")
            print("DRAW!")
            print_board()
            break


play()

File: four-connect.cpp
#include <iostream>
#include <vector>

const int ROWS = 6;
const int COLS = 7;
std::vector<std::vector<char>> arr(ROWS, std::vector<char>(COLS, '0'));

void printBoard() {
    for (int r = 0; r < ROWS; ++r) {
        for (int c = 0; c < COLS; ++c) {
            char endChar = (COLS - 1 == c) ? '\n' : ' ';
            std::cout << arr[r][c] << " " << endChar;
        }
    }
}

bool placeCoin(int col, char coin) {
    int row = 0;
    if (arr[row][col] != '0') {
        std::cout << "Column is full, try again\n";
        return false;
    }

    while (arr[row][col] == '0') {
        if (row + 1 == ROWS || arr[row + 1][col] != '0') {
            arr[row][col] = coin;
            break;
        }
        ++row;
    }
    return true;
}

bool checkDraw() {
    for (int r = 0; r < ROWS; ++r) {
        for (int c = 0; c < COLS; ++c) {
            if (arr[r][c] == '0') {
                return false;
            }
        }
    }
    return true;
}

bool isOutOfBounds(int row, int col) {
    return row < 0 || col < 0 || col >= COLS || row >= ROWS;
}

bool checkIsWinner(char coin) {
    for (int r = 0; r < ROWS; ++r) {
        for (int c = 0; c < COLS; ++c) {
            if (arr[r][c] == coin) {
                std::vector<std::pair<int, int>> directions = {
                    {1, 0},    // Right
                    {-1, 0},   // Left
                    {0, 1},    // Up
                    {0, -1},   // Down
                    {1, 1},    // Up-Right
                    {-1, 1},   // Up-Left
                    {1, -1},   // Down-Right
                    {-1, -1}   // Down-Left
                };

                for (const auto& [dx, dy] : directions) {
                    int newR = dx + r;
                    int newC = dy + c;
                    int foundCoins = 1;

                    for (int i = 0; i < 3; ++i) {
                        if (isOutOfBounds(newR, newC)) {
                            break;
                        }

                        if (arr[newR][newC] == coin) {
                            ++foundCoins;
                        }

                        newR += dx;
                        newC += dy;
                    }

                    if (foundCoins == 4) {
                        return true;
                    }
                }
            }
        }
    }

    return false;
}

void play() {
    char currentTurn = 'R';
    while (true) {
        while (true) {
            system("cls");
            std::cout << "It's " << currentTurn << " Turn:\n";
            int columnInput;
            while (true) {
                std::cout << "Pick a column to place the coin (0-6):\n";
                printBoard();
                std::cin >> columnInput;

                if (isOutOfBounds(1, columnInput)) {
                    std::cout << "Column is out of bounds, please try again\n";
                } else {
                    break;
                }
            }

            bool isPlaced = placeCoin(columnInput, currentTurn);
            if (isPlaced) {
                currentTurn = (currentTurn == 'Y') ? 'R' : 'Y';
                break;
            }
        }

        if (checkIsWinner('Y')) {
            system("cls");
            std::cout << "Y won the game!\n";
            printBoard();
            break;
        }

        if (checkIsWinner('R')) {
            system("cls");
            std::cout << "R won the game!\n";
            printBoard();
            break;
        }

        if (checkDraw()) {
            system("cls");
            std::cout << "DRAW!\n";
            printBoard();
            break;
        }
    }
}

int main() {
    play();
    return 0;
}
